src/internal-links/suggestions-generator.js

```diff
@@ -0,0 +1,151 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { getPrompt } from '@adobe/spacecat-shared-utils';
+import { FirefallClient } from '@adobe/spacecat-shared-gpt-client';
+import { getScrapedDataForSiteId, sleep } from '../support/utils.js';
+
+export const generateSuggestionData = async (finalUrl, auditData, context, site) => {
+  const { dataAccess, log } = context;
+  const { Configuration } = dataAccess;
+
+  if (auditData.auditResult.success === false) {
+    log.info('Audit failed, skipping suggestions generation');
+    return { ...auditData };
+  }
+
+  const configuration = await Configuration.findLatest();
+  if (!configuration.isHandlerEnabledForSite('broken-internal-links-auto-suggest', site)) {
+    log.info('Auto-suggest is disabled for site');
+    return { ...auditData };
+  }
+
+  log.info(`Generating suggestions for site ${finalUrl}`);
+
+  const firefallClient = FirefallClient.createFrom(context);
+  const firefallOptions = { responseFormat: 'json_object' };
+  const BATCH_SIZE = 300;
+
+  const data = await getScrapedDataForSiteId(site, context);
+  const { siteData, headerLinks } = data;
+  const totalBatches = Math.ceil(siteData.length / BATCH_SIZE);
+  const dataBatches = Array.from(
+    { length: totalBatches },
+    (_, i) => siteData.slice(i * BATCH_SIZE, (i + 1) * BATCH_SIZE),
+  );
+
+  log.info(`Processing ${siteData.length} alternative URLs in ${totalBatches} batches of ${BATCH_SIZE}...`);
+
+  const processBatch = async (batch, urlTo) => {
+    try {
+      const requestBody = await getPrompt({ alternative_urls: batch, broken_url: urlTo }, 'broken-backlinks', log);
+      await sleep(1000);
+      const response = await firefallClient.fetchChatCompletion(requestBody, firefallOptions);
+
+      if (response.choices?.length >= 1 && response.choices[0].finish_reason !== 'stop') {
+        log.error(`No suggestions found for ${urlTo}`);
+        return null;
+      }
+
+      return JSON.parse(response.choices[0].message.content);
+    } catch (error) {
+      log.error(`Batch processing error: ${error.message}`);
+      return null;
+    }
+  };
+
+  const processLink = async (link, headerSuggestions) => {
+    log.info(`Processing link: ${link.url_to}`);
+    const suggestions = [];
+    for (const batch of dataBatches) {
+      // eslint-disable-next-line no-await-in-loop
+      const result = await processBatch(batch, link.url_to);
+      if (result) {
+        suggestions.push(result);
+      }
+    }
+
+    if (totalBatches > 1) {
```

Comment:
I would extract this part of the code or explicitly pass the parameter.

Right now, the function's behavior depends on a variable from the context, introducing **temporal coupling**.

Reply:
duplicate